Return 404 page if locale is unavailable or no header sent

robhudson opened the pull request:
## One-line summary

This changes the default of unsupported locales from redirect to `/en-US/` to a 404 with a locale selection page.

## Testing

The redirect to best locale logic depends on the `Accept-language` header. If yours is set to an English locale it will be hard to trigger the 404 locale selection page. So this is what I have done:

1. Set your preferred language in your browser settings to remove English and pick an uncommonly supported locale. I chose `ast` since it was near the top.
2. Navigate to localhost:8000 or localhost:8000/firefox/new/

You can also use `curl` to test responses with and without an accept language header like so:

1. `curl -IL localhost:8000/ -H "Accept-language: ast"`
2. `curl -IL localhost:8000/`

stevejalim commented:
Heya @robhudson - just tried to test drive this but bumped up against a couple of niggles and  - maybe I need ot set something else?
1. We need DEBUG=True in .env to trigger the 404 page
2. I get a 500 when using the langauge code `ch` - do I need to do anything else to bootstrap things?
```
[0] Traceback (most recent call last):
[0]   File "/Users/steve/.pyenv/versions/3.9.10/lib/python3.9/wsgiref/handlers.py", line 137, in run
[0]     self.result = application(self.environ, self.start_response)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/sentry_sdk/integrations/django/__init__.py", line 127, in sentry_patched_wsgi_handler
[0]     return SentryWsgiMiddleware(bound_old_app, use_x_forwarded_for)(
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/sentry_sdk/integrations/wsgi.py", line 139, in __call__
[0]     reraise(*_capture_exception(hub))
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/sentry_sdk/_compat.py", line 54, in reraise
[0]     raise value
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/sentry_sdk/integrations/wsgi.py", line 132, in __call__
[0]     rv = self.app(
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django/core/handlers/wsgi.py", line 133, in __call__
[0]     response = self.get_response(request)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/sentry_sdk/integrations/django/__init__.py", line 391, in sentry_patched_get_response
[0]     rv = old_get_response(self, request)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django/core/handlers/base.py", line 130, in get_response
[0]     response = self._middleware_chain(request)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django/core/handlers/exception.py", line 49, in inner
[0]     response = response_for_exception(request, exc)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django/core/handlers/exception.py", line 114, in response_for_exception
[0]     response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django/core/handlers/exception.py", line 153, in handle_uncaught_exception
[0]     return callback(request)
[0]   File "/Users/steve/Code/bedrock/bedrock/base/views.py", line 169, in server_error_view
[0]     return l10n_utils.render(request, template_name, ftl_files=["500"], status=500)
[0]   File "/Users/steve/Code/bedrock/lib/l10n_utils/__init__.py", line 158, in render
[0]     return redirect_to_best_locale(request, translations)
[0]   File "/Users/steve/Code/bedrock/lib/l10n_utils/__init__.py", line 57, in redirect_to_best_locale
[0]     return locale_selection(request, translations)
[0]   File "/Users/steve/Code/bedrock/lib/l10n_utils/__init__.py", line 74, in locale_selection
[0]     return django_render(request, "404-locale.html", context, status=404)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/sentry_sdk/integrations/django/templates.py", line 95, in render
[0]     return real_render(request, template_name, context, *args, **kwargs)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django/shortcuts.py", line 19, in render
[0]     content = loader.render_to_string(template_name, context, request, using=using)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django/template/loader.py", line 62, in render_to_string
[0]     return template.render(context, request)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django_jinja/backend.py", line 59, in render
[0]     return mark_safe(self._process_template(self.template.render, context, request))
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/django_jinja/backend.py", line 105, in _process_template
[0]     return handler(context)
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/jinja2/environment.py", line 1291, in render
[0]     self.environment.handle_exception()
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/jinja2/environment.py", line 925, in handle_exception
[0]     raise rewrite_traceback_stack(source=source)
[0]   File "/Users/steve/Code/bedrock/bedrock/base/templates/404-locale.html", line 7, in top-level template code
[0]     {% extends "base-protocol-mozilla.html" %}
[0]   File "/Users/steve/Code/bedrock/bedrock/base/templates/base-protocol-mozilla.html", line 7, in top-level template code
[0]     {% extends 'base-protocol.html' %}
[0]   File "/Users/steve/Code/bedrock/bedrock/base/templates/base-protocol.html", line 100, in top-level template code
[0]     {% block content %}{% endblock %}
[0]   File "/Users/steve/Code/bedrock/bedrock/base/templates/404-locale.html", line 31, in block 'content'
[0]     <li lang="{{ locale }}"><a href="/{{ locale }}{{ request.path_info }}" title="Browse {{ request.path_info }} in the {{ languages[locale]['English'] }} language">{{ languages[locale]["native"] }}</a></li>
[0]   File "/Users/steve/.pyenv/versions/bedrock39/lib/python3.9/site-packages/jinja2/environment.py", line 455, in getitem
[0]     return obj[argument]
[0] jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'am'
```

stevejalim commented on bedrock/base/templates/404-locale.html:
```diff
@@ -0,0 +1,37 @@
+{% endblock %}
+
+{% block canonical_urls %}
+  {# Should 404s have canonical URLs? #}
```
I think @a-kyne will know the best answer to that question, but I'd say maybe not, as it's not a page we generally drive traffic _to_

stevejalim commented on bedrock/careers/tests/test_views.py:
```diff
@@ -15,7 +15,7 @@
 class PositionTests(TestCase):
     def test_context(self):
-        response = self.client.get(reverse("careers.listings"), follow=True)
+        response = self.client.get(reverse("careers.listings"), follow=True, HTTP_ACCEPT_LANGUAGE="en")
```
Side-thought: to avoid the locale header boilerplate, would it be worth using a [custom Client class](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#customizing-the-test-client)? It looks like we could make it a [default header](https://github.com/django/django/blob/db588d4f0edde5095f8975be1da4e693268b86b3/django/test/client.py#L383)

robhudson replied:
Possibly? But I think it's safe to assume an accept-language header for tests since that's what browsers will be doing. And we could override it explicitly to test the header-less flow. I can look into that.

stevejalim commented on lib/l10n_utils/tests/test_files/templates/404-locale.html:
```diff
@@ -0,0 +1,6 @@
+ License, v. 2.0. If a copy of the MPL was not distributed with this
+ file, You can obtain one at https://mozilla.org/MPL/2.0/.
+#}
+
```
I'm assuming this file just needs to exist to make the tests work happily in isolation, yep?

robhudson replied:
yep!

stevejalim commented on tests/redirects/base.py:
```diff
@@ -131,8 +131,11 @@ def assert_valid_url(
     kwargs = {"allow_redirects": follow_redirects}
-    if req_headers:
+    if req_headers is None:
+        kwargs["headers"] = {"Accept-language": "en"}
```
Was this added to avoid having to patch a ton of existing tests?

robhudson replied:
Yes, for the redirects it made sense to me as a default.

stevejalim commented on tests/redirects/map_globalconf.py:
```diff
@@ -780,7 +780,7 @@
         # bug 1248393
         url_test("/de/about/legal/impressum/", status_code=requests.codes.ok),
-        url_test("/{en-US,fr,ja}/about/legal/impressum/", "/de/about/legal/impressum/", status_code=requests.codes.found),
+        url_test("/{en-US,fr,ja}/about/legal/impressum/", status_code=requests.codes.not_found),
```
I know that the German Impressum page is an important one that only exists in `de`. I can't currently run the code to check, but my assumption here is that the 404 page would have a link to `/de/about/legal/impressum/` yeah?

robhudson replied:
Yep! 👍

stevejalim reviewed (commented):
Liking this a lot! Looks clean and smart - thanks for the work!

Once I get it running locally (see other comment) I'll be happy to test drive it, too

maureenlholland commented:
@stevejalim 

re:
```
[0] jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'am'
```
I definitely hit a similar error, but can't recall the exact setup I was using at the time. 
I've added `Dev=False` in my env settings since we don't want all locales active for every page. (according to prod mode docs "you may want DEV=True if you want the site to act more like www-dev.allizom.org in that all feature switches are On and all locales are active for every page.").

I removed all languages from my browser language settings to get unblocked and that worked just fine. I re-added AST to test, and it does serve the expected template for http://localhost:8000/ast/firefox/, if you want to give that specific combo a try.

stevejalim commented:
@maureenlholland Thanks for this

> Looks like Dev=False is the key

However, I guess it also means we need to address Dev=True, else we won't be able to test/use this behaviour on www-dev.allizom.org, yeah?

maureenlholland commented:
@stevejalim good point for staging tests/usage

stevejalim commented:
Got it running locally and the mechanism is nice ✨ 👍 

Couple of things we should prob consider though:

1) We should make this work with `DEV=True`, so we don't break www-dev.allizom.org - even if that means that the language selection is just the small number of dev langs. Is that a world of hurt to make happen?

2) When I first saw the pick-a-language page, I was confused because it didn't appear connected to the page I was trying to reach until I hovered over the locale link options:

<img width="1433" alt="Screenshot 2022-06-14 at 13 24 31" src="https://user-images.githubusercontent.com/101457/173576688-1cfaed55-2eb4-4281-82ac-0b940c9fa536.png">

I think it's worth showing the user that although they've 404ed, they haven't just fallen into a black hole - eg

<img width="1433" alt="404_1" src="https://user-images.githubusercontent.com/101457/173582000-91a224ea-21e8-4ade-8b28-152d2a4da5bd.png">
<img width="1433" alt="404_2" src="https://user-images.githubusercontent.com/101457/173582020-5f4d0d74-b5f3-4072-a736-5a2253494770.png">

(Feel free to call on a real UXer, not me 😄 )

alexgibson commented:
> We should make this work with DEV=True, so we don't break www-dev.allizom.org - even if that means that the language selection is just the small number of dev langs. Is that a world of hurt to make happen?

Apologies if this has already been discussed: it's probably worth calling out that translators use www-dev to work on their translations. With that in mind, it's likely useful if we didn't show the 404 page on dev if a locale is untranslated, so that localizers can continue to preview their work?

a-kyne commented:
What do we think about having the not-translated 404 response page say, "We're sorry, that web page has not yet been translated by our volunteers. It is available in the following languages:"

a-kyne commented:
FYI, as long as the error page is using the 404 status code, the SEO requirement is satisfied. For the usability component, let's check with Elise.

a-kyne commented:
BTW a 404 error page does not need a canonical URL. The status code ensures that no one is going to try to index it.

robhudson commented:
Thanks all!

- I'll checkout the `Dev=True` case. I have `Dev=False` locally and hadn't tested that and agree it should work in both states.
- And agree with everything here about the UI issues on the 404 page. I can help find consensus on best approach with help from FE/UI.

maureenlholland commented:
@ehot what's the UX perspective here? This is the 404 page served when a user tries to access a page not available in their locale (as determined from path or browser language settings).

Currently there is a simple "Choose your language" heading and a list of available languages and a bottom section with the following copy:
heading "Want to contribute?"
paragraph with link "Learn more about Mozilla’s volunteer-driven [localization community](https://wiki.mozilla.org/L10n:Contribute)

These suggestions definitely sound like an improvement to add under "Choose your language" for context:
- "We're sorry, that web page has not yet been translated by our volunteers. It is available in the following languages:"
- "We're sorry, that web page has not yet been [translated by our community](https://community.mozilla.org/en/activities/localize-mozilla/). It is available in the following languages:"

Any copy adjustments you'd like to see? Should we make the path to start contributing more explicit? I'd vote to keep the wikimo link over the community.mozilla one, otherwise, no strong preferences.

robhudson commented:
I fixed the 500 when `DEV=True` error. The issue was that the `www-l10n` folder has locales in it that don't match what is in `product_details`, causing a key error. I added a conditional check to the template to skip these, but it makes me wonder if those should be removed from the `www-l10n` repo if we can't support them? Probably no harm to leave if there's a workaround though.

maureenlholland commented:
This is current state of front-end
<img width="789" alt="locale-404" src="https://user-images.githubusercontent.com/19650432/174082609-704b96bc-ad8f-446c-aed9-f6044d17976e.png">

ehot commented:
![translation-flow](https://user-images.githubusercontent.com/2040610/174186569-0e00a1f5-7e17-46e4-9c13-44e052a24209.png)
Sorry for my delayed / late reply. 

I think we can alter the tone so that no one has to say sorry (us / community / etc.). It's just work that we haven't gotten to and everyone's human. Instead, I think offering them a chance to contribute in their native language would be really kind and inclusive. I'm afraid that if a person is reading something in their native language and then they get a 404 page in english, it may be jarring. Also, assuming we can get permission, maybe we can start leveraging the illustration styles for mozilla including this one I altered to have a conversation between the dino and a person.

maureenlholland commented:
Thanks @ehot! I will update text to match what you've got here and remove the bottom contribute section. The dino illustration is lovely, who do we have to ask for permission? I can add an issue to upgrade the image here when that's all set, but wouldn't consider it a blocker for this version.

It would be great to have this page shown in the language the user tried to access their originally requested page in, but we will also have to consider the case of: 
- user is directed to locale-404 because page is not available in their requested language
- locale-404 is not available in their requested language
- user can't be redirected to locale-404 or they'll just loop, so we serve an EN fallback locale-404 page

@robhudson I'll set up the FTL file in a separate commit. Please feel free to revert if this isn't technically possible or the added logic is a bit more than we want to manage in this version.

maureenlholland commented on bedrock/base/templates/404-locale.html:
```diff
@@ -24,14 +24,14 @@
     <ul>
       {% for locale in available_locales if locale in languages %}
-        <li lang="{{ locale }}"><a href="/{{ locale }}{{ request.path_info }}" title="Browse {{ request.path_info }} in the {{ languages[locale]['English'] }} language">{{ languages[locale]["native"] }}</a></li>
+        <li lang="{{ locale }}"><a href="/{{ locale }}{{ request.path_info }}" title="{{ ftl('not-found-locale-link-title', requested_page=request.path_info, link_language=languages[locale]['English']) }}">{{ languages[locale]["native"] }}</a></li>
```
link_language probably should be [locale]['native'] if we keep translations

maureenlholland commented:
Apologies @robhudson, I may have broken everything 🙈

You can get a basic view of the translations on http://localhost:8000/en-US/404-locale/ with `DEBUG=True`, but I have no idea how to mock up the type of context needed for a better test of the translated page (request path, available languages).

Also no idea how to avoid the crashing loop on non-debug environment. Hoping you have some ideas.

ehot commented:
I'm sorry; I'm late again in replying. I'm gaining permission as we speak. These illustrations were made in-house, for use by mozilla, but for a different concept, so I'm just double checking.

ehot commented:
I'm sorry; I'm late again in replying. I'm gaining permission as we speak. These illustrations were made in-house, for use by mozilla, but for a different concept, so I'm just double checking.

stevejalim commented:
@robhudson Not sure if this is ready for a re-review, but if it is, just let me know. Happy to.

stevejalim reviewed (commented):
`r+`! Extra `+++` for handling so many edges and corners of the various situations
